### DISCUSSION TITLE
Setting an SSH preference after installation

### DISCUSSION THREAD

**Discussion by Sophismata** [git-for-windows/git#3451] [Q&A] [answered]
When installing Git For Windows via the installer, you have an option to choose if Git For Windows should use its built-in SSH provider, or an external SSH provider.

There's no clear setting in `git config` to switch this on or off. How can this preference be configured post-installation, and how can it be configured in the portable version of Git For Windows?

**Comment by dscho** [comment, marked as answer]
[This](https://github.com/git-for-windows/build-extra/blob/7977f90a119a08d144fa885071eb59dc1b3474eb/installer/install.iss#L3353-L3359) is the "business logic" of that toggle. It deletes the files originating from the `openssh` package that were bundled with Git for Windows.

There is no easy way to replicate the exact logic in Portable Git. But you can emulate to a certain extent it by deleting the relevant files from `/usr/bin/`: `rm /usr/bin/{scp,sftp,ssh*}.exe`.

**Reply by Sophismata** [reply]
Thank you. I assume this means Get For Windows simply defers to the first `openssh` executable found on the PATH? (And by removing the bundled `openssh` it will naturally fall back to an external executable, if found.)

**Reply by Sophismata** [reply]
I've also added an article to the wiki — [Using an external OpenSSH client](https://github.com/git-for-windows/git/wiki/Using-an-external-OpenSSH-client). Please feel free to remove it if doing so was inappropriate.

**Reply by dscho** [reply]
> I assume this means Get For Windows simply deferes to the first `openssh` executable found on the PATH?

Yes, that's correct!

> I've also added an article to the wiki

Fantastic, thank you!

**Reply by Okeanos** [reply]
The reason why it was implemented as such is that Git for Windows manipulates the `$PATH` in such a way that prepending something else to override the bundled OpenSSH binaries isn't easily done.

**Reply by Sophismata** [reply]
Thanks for the info. It makes sense to me, and in spite of the, um, 'quick win' the feature is working well so far. Windows doesn't have the full SSH toolkit but it's nice to be able to have the ssh-agent run as a Windows service & have consistent private key management across the whole operating environment.

Ultimately if the feature is kept a config setting would be the ideal way to manage it. But I don't imagine there's much demand or need presently, particularly as WSL does much of the heavy lifting for SSH usage in Windows right now, in ops & development.

Maybe as Windows Server Core becomes more common [OpenSSH in Windows](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_overview) will see more use.

**Reply by dscho** [reply]
> Ultimately if the feature is kept a config setting would be the ideal way to manage it.

You can always do that yourself: https://git-scm.com/docs/git-config#Documentation/git-config.txt-coresshCommand

**Reply by Okeanos** [reply]
> > Ultimately if the feature is kept a config setting would be the ideal way to manage it.
> 
> You can always do that yourself: https://git-scm.com/docs/git-config#Documentation/git-config.txt-coresshCommand

As an addendum: overriding `GIT_SSH_COMMAND` was already possible before this change, however, only touches `git` execution – the current 'quick win'-fix (not installing bundled OpenSSH) also forces _other_ tools shipped alongside Git for Windows as well as tools that are user installed, to use a single, predictable OpenSSH version. `GIT_SSH_COMMAND` isn't honoured by say … KeePassXC or whatever.
It might be worthwhile to investigate whether the `$PATH` manipulation done by Git for Windows can be customized in such a way to allow user pathing (via Windows Environment Settings and not `.bashrc`) that overrides bundled software. That would effectively obsolete the current 'quick fix' and allow switching between external & bundled OpenSSH on demand.

**Reply by Sophismata** [reply]
Thank you both.

**Reply by keviinalexis1430-star** [reply]
> [This](https://github.com/git-for-windows/build-extra/blob/7977f90a119a08d144fa885071eb59dc1b3474eb/installer/install.iss#L3353-L3359) is the "business logic" of that toggle. It deletes the files originating from the `openssh` package that were bundled with Git for Windows.
> 
> There is no easy way to replicate the exact logic in Portable Git. But you can emulate to a certain extent it by deleting the relevant files from `/usr/bin/`: `rm /usr/bin/{scp,sftp,ssh*}.exe`.
